File: agent-js/src/lifecycle-failures.test.ts
import { LanguageModelError } from "@hoangvvo/llm-sdk";
import { MockLanguageModel } from "@hoangvvo/llm-sdk/test";
import test, { suite, type TestContext } from "node:test";
import { Agent } from "./agent.ts";
import { AgentInitError, AgentLanguageModelError } from "./errors.ts";
import { RunSession } from "./run.ts";

suite("agent lifecycle failures", () => {
  test("closes toolkit sessions that initialized before another toolkit failed", async (t: TestContext) => {
    let closeCalls = 0;
    const initFailure = new Error("later toolkit failed");
    const cleanupFailure = new Error("cleanup failed");
    const model = new MockLanguageModel();

    await t.assert.rejects(
      () =>
        RunSession.create({
          name: "test_agent",
          model,
          context: {},
          toolkits: [
            {
              createSession: () =>
                Promise.resolve({
                  getSystemPrompt: () => undefined,
                  getTools: () => [],
                  close: () => {
                    closeCalls++;
                    throw cleanupFailure;
                  },
                }),
            },
            {
              createSession: () =>
                Promise.resolve({
                  getSystemPrompt: () => undefined,
                  getTools: () => [],
                  close: () => {
                    closeCalls++;
                    return Promise.resolve();
                  },
                }),
            },
            { createSession: () => Promise.reject(initFailure) },
          ],
        }),
      (error: unknown) => {
        t.assert.strictEqual(error instanceof AgentInitError, true);
        t.assert.strictEqual((error as AgentInitError).cause, initFailure);
        return true;
      },
    );
    t.assert.strictEqual(closeCalls, 2);
  });

  test("attempts every toolkit close and reports cleanup failure", async (t: TestContext) => {
    const cleanupFailure = new Error("cleanup failed");
    let successfulCloseCalls = 0;
    const model = new MockLanguageModel();
    const session = await RunSession.create({
      name: "test_agent",
      model,
      context: {},
      toolkits: [
        {
          createSession: () =>
            Promise.resolve({
              getSystemPrompt: () => undefined,
              getTools: () => [],
              close: () => {
                throw cleanupFailure;
              },
            }),
        },
        {
          createSession: () =>
            Promise.resolve({
              getSystemPrompt: () => undefined,
              getTools: () => [],
              close: () => {
                successfulCloseCalls++;
                return Promise.resolve();
              },
            }),
        },
      ],
    });

    await t.assert.rejects(() => session.close(), cleanupFailure);
    t.assert.strictEqual(successfulCloseCalls, 1);
  });

  test("preserves the model error when cleanup also fails", async (t: TestContext) => {
    const modelFailure = new LanguageModelError("generation failed");
    const cleanupFailure = new Error("cleanup failed");
    const model = new MockLanguageModel();
    model.enqueueGenerateResult({ error: modelFailure });
    const agent = new Agent({
      name: "test_agent",
      model,
      toolkits: [
        {
          createSession: () =>
            Promise.resolve({
              getSystemPrompt: () => undefined,
              getTools: () => [],
              close: () => Promise.reject(cleanupFailure),
            }),
        },
      ],
    });

    await t.assert.rejects(
      () =>
        agent.run({
          context: {},
          input: [
            {
              type: "message",
              role: "user",
              content: [{ type: "text", text: "Hello" }],
            },
          ],
        }),
      (error: unknown) => {
        t.assert.strictEqual(error instanceof AgentLanguageModelError, true);
        t.assert.strictEqual(
          (error as AgentLanguageModelError).cause,
          modelFailure,
        );
        return true;
      },
    );
  });
});

File: agent-go/lifecycle_failures_test.go
package llmagent_test

import (
	"context"
	"errors"
	"testing"

	llmagent "github.com/hoangvvo/llm-sdk/agent-go"
	llmsdk "github.com/hoangvvo/llm-sdk/sdk-go"
	"github.com/hoangvvo/llm-sdk/sdk-go/llmsdktest"
)

func TestRunSession_ClosesPartialInitializationOnToolkitFailure(t *testing.T) {
	initialized := &mockToolkitSession[struct{}]{}
	initFailure := errors.New("second toolkit failed")
	model := llmsdktest.NewMockLanguageModel()
	_, err := llmagent.NewRunSession(
		t.Context(),
		&llmagent.AgentParams[struct{}]{
			Name:  "test_agent",
			Model: model,
			Toolkits: []llmagent.Toolkit[struct{}]{
				&mockToolkit[struct{}]{createFn: func(context.Context, struct{}) (llmagent.ToolkitSession[struct{}], error) {
					return initialized, nil
				}},
				&mockToolkit[struct{}]{createFn: func(context.Context, struct{}) (llmagent.ToolkitSession[struct{}], error) {
					return nil, initFailure
				}},
			},
			MaxTurns: 10,
		},
		struct{}{},
	)
	if !errors.Is(err, initFailure) {
		t.Fatalf("expected initialization failure, got %v", err)
	}
	if initialized.closeCalls != 1 {
		t.Fatalf("expected initialized toolkit cleanup, got %d", initialized.closeCalls)
	}
}

func TestRunSession_CloseAttemptsEveryToolkitAndReportsFailure(t *testing.T) {
	cleanupFailure := errors.New("cleanup failed")
	failing := &mockToolkitSession[struct{}]{closeErr: cleanupFailure}
	successful := &mockToolkitSession[struct{}]{}
	model := llmsdktest.NewMockLanguageModel()
	session := mustNewRunSession(
		t,
		&llmagent.AgentParams[struct{}]{
			Name:  "test_agent",
			Model: model,
			Toolkits: []llmagent.Toolkit[struct{}]{
				&mockToolkit[struct{}]{createFn: func(context.Context, struct{}) (llmagent.ToolkitSession[struct{}], error) {
					return failing, nil
				}},
				&mockToolkit[struct{}]{createFn: func(context.Context, struct{}) (llmagent.ToolkitSession[struct{}], error) {
					return successful, nil
				}},
			},
			MaxTurns: 10,
		},
		struct{}{},
	)

	err := session.Close(t.Context())
	if !errors.Is(err, cleanupFailure) {
		t.Fatalf("expected cleanup failure, got %v", err)
	}
	if failing.closeCalls != 1 || successful.closeCalls != 1 {
		t.Fatalf("expected every toolkit to close: failing=%d successful=%d", failing.closeCalls, successful.closeCalls)
	}
}

func TestAgentRun_PreservesModelErrorWhenCleanupFails(t *testing.T) {
	modelFailure := llmsdk.NewInvalidInputError("generation failed")
	cleanupFailure := errors.New("cleanup failed")
	model := llmsdktest.NewMockLanguageModel()
	model.EnqueueGenerateResult(llmsdktest.NewMockGenerateResultError(modelFailure))
	toolkitSession := &mockToolkitSession[struct{}]{closeErr: cleanupFailure}
	agent := llmagent.NewAgent(
		"test_agent",
		model,
		llmagent.WithToolkits[struct{}](&mockToolkit[struct{}]{createFn: func(context.Context, struct{}) (llmagent.ToolkitSession[struct{}], error) {
			return toolkitSession, nil
		}}),
	)

	_, err := agent.Run(t.Context(), llmagent.AgentRequest[struct{}]{
		Context: struct{}{},
		Input: []llmagent.AgentItem{
			llmagent.NewAgentItemMessage(llmsdk.NewUserMessage(llmsdk.NewTextPart("Hello"))),
		},
	})
	if !errors.Is(err, modelFailure) || errors.Is(err, cleanupFailure) {
		t.Fatalf("expected primary model error, got %v", err)
	}
}
